File: src/dstf/simulators/python/run_test.py
"""Framework-shipped Python driver for a single test (D80).

Invoked by PythonRunner as:
    python run_test.py <user_file> <stop_time> <tolerance> <result_path>

The user file must define:
    simulate(stop_time: float, tolerance: float) -> dict

returning a dict with:
    "time":      list[float] (monotonically non-decreasing)
    "variables": dict[str, list[float]]  (each same length as "time")

Result file is JSON:
    {"success": true,  "time": [...], "variables": {...}}
or on failure:
    {"success": false, "error": "...", "time": [], "variables": {}}

Exceptions are caught so the framework always gets a structured result
rather than a process-level crash.
"""

from __future__ import annotations

import importlib.util
import json
import sys
import traceback
from pathlib import Path


def _load_user_module(user_file: Path):
    # Allow sibling imports from the user's file directory.
    parent = str(user_file.parent.resolve())
    if parent not in sys.path:
        sys.path.insert(0, parent)
    spec = importlib.util.spec_from_file_location("_user_test", user_file)
    if spec is None or spec.loader is None:
        raise ImportError(f"Cannot load {user_file}")
    module = importlib.util.module_from_spec(spec)
    spec.loader.exec_module(module)
    return module


def _write_failure(result_path: Path, message: str) -> None:
    result_path.write_text(
        json.dumps(
            {
                "success": False,
                "error": message,
                "time": [],
                "variables": {},
            }
        ),
        encoding="utf-8",
    )


def main(argv: list[str]) -> int:
    if len(argv) != 4:
        print(
            "Usage: run_test.py <user_file> <stop_time> <tolerance> <result_path>",
            file=sys.stderr,
        )
        return 2

    user_file = Path(argv[0])
    try:
        stop_time = float(argv[1])
        tolerance = float(argv[2])
    except ValueError as exc:
        print(f"Invalid numeric arg: {exc}", file=sys.stderr)
        return 2
    result_path = Path(argv[3])

    try:
        module = _load_user_module(user_file)
        if not hasattr(module, "simulate"):
            raise AttributeError(
                f"{user_file.name} must define simulate(stop_time, tolerance)"
            )
        payload = module.simulate(stop_time, tolerance)
        if (
            not isinstance(payload, dict)
            or "time" not in payload
            or "variables" not in payload
        ):
            raise ValueError(
                "simulate() must return a dict with 'time' and 'variables' keys"
            )
        if not isinstance(payload["variables"], dict):
            raise ValueError("'variables' must be a dict of name -> list[float]")

        # review 2026-07-06 (finding 29): validate value TYPES, not just list
        # lengths — a simulate() returning strings used to pass the run and
        # then blow up np.asarray during read_results, aborting every test's
        # read phase. Fail here, cleanly, naming the offending variable.
        raw_time = list(payload["time"])
        try:
            time_list = [float(v) for v in raw_time]
        except (TypeError, ValueError) as exc:
            raise ValueError(f"'time' contains a non-numeric value: {exc}") from exc
        variables = {}
        for k, v in payload["variables"].items():
            name = str(k)
            raw_values = list(v)
            if len(raw_values) != len(time_list):
                raise ValueError(
                    f"variable '{name}' has {len(raw_values)} samples; "
                    f"expected {len(time_list)} to match 'time'"
                )
            try:
                variables[name] = [float(x) for x in raw_values]
            except (TypeError, ValueError) as exc:
                raise ValueError(
                    f"variable '{name}' contains a non-numeric value: {exc}"
                ) from exc

        result_path.write_text(
            json.dumps(
                {
                    "success": True,
                    "time": time_list,
                    "variables": variables,
                }
            ),
            encoding="utf-8",
        )
        print(f"OK: {len(variables)} variables, {len(time_list)} time points")
        return 0
    except BaseException:
        tb = traceback.format_exc()
        print(f"FAIL:\n{tb}", file=sys.stderr)
        try:
            _write_failure(result_path, tb)
        except Exception:
            return 2
        return 1


if __name__ == "__main__":
    sys.exit(main(sys.argv[1:]))

File: src/dstf/simulators/julia/run_test.jl
# Framework-generated Julia driver for a single test.
#
# Invoked by JuliaRunner as:
#   julia --project=<examples_project> run_test.jl <user_file> <stop_time> <tol> <result_path>
#
# The user file must export `build_mtk_system()` returning a NamedTuple with:
#   sys :: ODESystem   — structurally simplified
#   u0  :: Vector      — initial conditions (pairs of Symbol => value)
#   ps  :: Vector      — parameter values    (pairs of Symbol => value)
#
# Result file is JSON:
#   { "success": bool,
#     "time": [...],
#     "variables": {"name": [...], ...},
#     "error": "..." (only if success == false) }
#
# The wrapper intentionally catches exceptions so the framework always
# gets a structured result (success=false + error) rather than a
# process-level crash.

using Printf

const USER_FILE   = ARGS[1]
const STOP_TIME   = parse(Float64, ARGS[2])
const TOLERANCE   = parse(Float64, ARGS[3])
const RESULT_PATH = ARGS[4]

# review 2026-07-06 (finding 75): the failure JSON is hand-built (JSON3 may
# not have loaded when the failure happens). Julia's repr()-style escaping
# is NOT JSON (`\$` is invalid) — real error detail was lost to parse
# errors. Minimal manual JSON string escaper instead.
function json_escape(s::AbstractString)
    buf = IOBuffer()
    for c in s
        if c == '\\'
            write(buf, "\\\\")
        elseif c == '"'
            write(buf, "\\\"")
        elseif c == '\n'
            write(buf, "\\n")
        elseif c == '\r'
            write(buf, "\\r")
        elseif c == '\t'
            write(buf, "\\t")
        elseif c < ' '
            write(buf, "\\u" * string(UInt32(c), base=16, pad=4))
        else
            write(buf, c)
        end
    end
    return String(take!(buf))
end

function write_failure(msg)
    open(RESULT_PATH, "w") do io
        write(io, string("{\"success\": false, \"error\": \"",
                         json_escape(string(msg)),
                         "\", \"time\": [], \"variables\": {}}"))
    end
end

try
    import ModelingToolkit
    import OrdinaryDiffEq
    import JSON3

    # Load the user file. It should export `build_mtk_system()`.
    include(USER_FILE)
    isdefined(@__MODULE__, :build_mtk_system) || error(
        "user file $(USER_FILE) must export build_mtk_system()")

    nt = build_mtk_system()
    haskey(nt, :sys) || error("build_mtk_system() must return NamedTuple with :sys")

    u0 = haskey(nt, :u0) ? nt.u0 : []
    ps = haskey(nt, :ps) ? nt.ps : []

    prob = ModelingToolkit.ODEProblem(nt.sys, u0, (0.0, STOP_TIME), ps)
    sol = OrdinaryDiffEq.solve(prob, OrdinaryDiffEq.Tsit5(); reltol=TOLERANCE)

    # review 2026-07-06 (finding 4): a diverged / unstable / maxiters-
    # truncated solve must not be written as success — with --accept the
    # truncated trajectory would become the regression baseline. Check the
    # retcode AND that integration actually reached stop_time.
    if !OrdinaryDiffEq.SciMLBase.successful_retcode(sol)
        error("solver did not succeed: retcode=" * string(sol.retcode))
    end
    if isempty(sol.t) || abs(sol.t[end] - STOP_TIME) > 1e-6 * max(abs(STOP_TIME), 1.0)
        t_end = isempty(sol.t) ? "none" : string(sol.t[end])
        error("solve stopped at t=" * t_end * " before stop_time=" *
              string(STOP_TIME) * " (retcode=" * string(sol.retcode) * ")")
    end

    # Collect unknowns (state variables) AND observables (algebraic
    # expressions like `y ~ a*x1 + b*x2` that structural_simplify
    # promotes out of the unknowns list but are still materializable
    # from the solution via sol[expr]). Users authoring MTK tests
    # typically expect both to be accessible.
    unk = ModelingToolkit.unknowns(nt.sys)
    obs_eqs = ModelingToolkit.observed(nt.sys)
    variables = Dict{String, Vector{Float64}}()
    for u in unk
        name = string(u)
        stripped = endswith(name, "(t)") ? name[1:end-3] : name
        variables[stripped] = Float64.(sol[u])
    end
    for eq in obs_eqs
        lhs = eq.lhs
        name = string(lhs)
        stripped = endswith(name, "(t)") ? name[1:end-3] : name
        # Skip if already captured as an unknown (shouldn't happen but
        # defensive) or if the name is a helper symbol MTK introduced.
        if haskey(variables, stripped) || startswith(stripped, "ˍ")
            continue
        end
        try
            variables[stripped] = Float64.(sol[lhs])
        catch
            # Some observables can't be evaluated (e.g., dependent on
            # parameters only); skip silently rather than fail the test.
        end
    end

    payload = Dict(
        "success" => true,
        "time" => Float64.(sol.t),
        "variables" => variables,
    )
    open(RESULT_PATH, "w") do io
        JSON3.write(io, payload)
    end

    @printf "OK: %d variables, %d time points\n" length(variables) length(sol.t)
catch e
    buf = IOBuffer()
    showerror(buf, e, catch_backtrace())
    msg = String(take!(buf))
    @printf stderr "FAIL: %s\n" msg
    try
        write_failure(msg)
    catch
        # If even writing the failure JSON fails, at least exit nonzero.
        exit(2)
    end
    exit(1)
end
